Products/CMFCore/indexing.py

```diff
@@ -0,0 +1,330 @@
+        catalog._indexObject(obj)
+
+    def reindex(self, obj, attributes=None, update_metadata=1):
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        catalog._reindexObject(obj, idxs=attributes, update_metadata=update_metadata)
+
+    def unindex(self, obj):
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        catalog._unindexObject(obj)
+
+    def begin(self):
+        pass
+
+    def commit(self):
+        pass
+
+    def abort(self):
+        pass
+
+    @staticmethod
+    def get_dispatcher(obj, name):
+        """ return named indexing method according on the used mixin class """
+        catalog = getToolByName(obj, 'portal_catalog', None)
+        if catalog is None:
+            return
+        attr = getattr(catalog, '_{0}'.format(name), None)
+        if attr is not None:
+            method = attr.im_func
+            return method
+        return
+
+
+def getQueue():
+    """ return a (thread-local) queue object, create one if necessary """
+    global localQueue
+    if localQueue is None:
+        localQueue = IndexQueue()
+    return localQueue
+
+
+def processQueue():
+    """ process the queue (for this thread) immediately """
+    queue = getQueue()
+    processed = 0
+    if queue.length() and not queue in processing:
+        debug('auto-flushing %d items: %r', queue.length(), queue.getState())
+        try:
+            processing.add(queue)
+            processed = queue.process()
+        finally:
+            processing.remove(queue)
+    return processed
+
+
+def wrap(obj):
+    """ the indexing key, i.e. the path to the object in the case of the
+        portal catalog, might have changed while the unindex operation was
+        delayed, for example due to renaming the object;  it was probably not
+        such a good idea to use a key that can change in the first place, but
+        to work around this a proxy object is used, which can provide the
+        original path;  of course, access to other attributes must still be
+        possible, since alternate indexers (i.e. solr etc) might use another
+        unique key, usually the object's uid;  also the inheritence tree
+        must match """
+    if getattr(aq_base(obj), 'getPhysicalPath', None) is None:
+        return obj
+
+    class PathWrapper(obj.__class__):
+
+        def __init__(self):
+            self.__dict__.update(dict(
+                context = obj,
+                path = obj.getPhysicalPath(),
+                REQUEST = getattr(obj, 'REQUEST', None)))
+
+        def __getattr__(self, name):
+            return getattr(aq_inner(self.context), name)
+
+        def __hash__(self):
+            return hash(self.context)   # make the wrapper transparent...
+
+        def getPhysicalPath(self):
+            return self.path
+
+    return PathWrapper().__of__(aq_parent(obj))
+
+
+@implementer(IIndexQueue)
+class IndexQueue(local):
+
+    def __init__(self):
+        self.queue = []
+        self.tmhook = None
```

Comment:
as far as I understand it sets only the register method w/o calling it immediately, so we do not benefit from prior instantiation.